### PR TITLE
Protocol 6.1

### PR DESCRIPTION
**Pull request by NichUK**
Protocol 6.1 PR as requested.
I'm sorry, it's a bit of a monster, but feel free to ask about anything that's not obvious, or point out some terrible coding. :)

I think it hits all the points from IQ's upgrade 6.0 to 6.1 release notes, but I haven't audited that.

Nich

### CODE REVIEW

**Comment by mathpaquette** [comment]
nice thanks for sharing. Looks promising. gimme some time to review.

**Comment by NichUK** [comment]
Happy to contribute.
As I said, there are definitely a couple of things that I rushed. The MarketSummaryMessage currently contains two copies of all the data which is obviously pointless. There's a comment in there with a couple of ideas. Interested to hear your thoughts when you have some time.

**Comment by NichUK** [comment]
I screwed up the FundamentalMessage with the protocol update. I didn't realise that they'd removed the reserved fields as well as adding new ones, and apparently didn't test that properly. I have a new commit that fixes it correctly.

**Comment by mathpaquette** [comment]
> I screwed up the FundamentalMessage with the protocol update. I didn't realise that they'd removed the reserved fields as well as adding new ones, and apparently didn't test that properly. I have a new commit that fixes it correctly.

did you run all the integration tests?

**Comment by mathpaquette** [comment]
dont worry, we gonna merge it at some point, Im just busy on other stuff..

**Comment by NichUK** [comment]
Sorry, I've been busy too. I did run integration tests, but of course, I updated the Fundamental Message test to what I thought was the new 6.1 protocol wire format... but I was wrong. :) 
I will have another look at everything very shortly.

**Comment by NichUK** [comment]
OK - I've been through and sorted the problem with FundamentalMessages and MarketSummaryMessages. It should be good to go now.

File: src/IQFeed.CSharpApiClient.Tests/Streaming/Level1/Messages/FundamentalMessageTests.cs [hunk 1]
```diff
@@ -14,11 +14,22 @@ public void Should_Parse_FundamentalMessage_Culture_Independent(string cultureNa
         {
             // Arrange
             TestHelper.SetThreadCulture(cultureName);
-            var message = "F,MSFT,5,65.6,28624000,97.9500,67.1400,97.9500,83.8300,,,,,,,,,,,1.49,,0.06,06,,MICROSOFT,MSFT,76.800,1.27,,159851.0,64527.0,03/31/2018,76073.0,7683198,,0.50 02/18/2003,0.50 03/29/1999,,,14,4,7372,24.82,1,21,05/10/2018,05/18/2017,05/10/2018,02/09/2018,85.5400,,,,,511210,,,,";
```
Comment (hunterfries) [inline src/IQFeed.CSharpApiClient.Tests/Streaming/Level1/Messages/FundamentalMessageTests.cs:17]: Is this more readable? I don't really think so, but open to discussion.

File: src/IQFeed.CSharpApiClient.Tests/Streaming/Level1/Messages/FundamentalMessageTests.cs [hunk 2]
```diff
@@ -14,11 +14,22 @@ public void Should_Parse_FundamentalMessage_Culture_Independent(string cultureNa
...
+            var message = "F,MSFT,5,65.6,28624000,97.9500,67.1400,97.9500,83.8300,,,,,,1.49," +
+                ",0.06,06,MICROSOFT,MSFT,76.800,1.27,,159851.0,64527.0,03/31/2018,76073.0," +
+                "7683198,0.50 02/18/2003,0.50 03/29/1999,14,4,7372,24.82,1,21,05/10/2018," +
+                "05/18/2017,05/10/2018,02/09/2018,85.5400,,,," +
+                ",511210,,,,09:00:00,17:00:00,USD,125000,HMUZ,0.1," +
+                "01/01/2016,TEST,1,";
 
             // Act
             var fundementalMessageParsed = FundamentalMessage.Parse(message);
-            var fundamentalMessage = new FundamentalMessage("MSFT", "5", 65.6d, 28624000, 97.95d, 67.14d, 97.95d, 83.83d, null, null, null, null, null, null, 1.49d, null, 0.06d, 6, "MICROSOFT", "MSFT", 76.8d, 1.27d, null, 159851, 64527, new DateTime(2018, 03, 31), 76073, 7683198, "0.50 02/18/2003", "0.50 03/29/1999", "14", 4, 7372, 24.82d, "1", "21", new DateTime(2018, 05, 10), new DateTime(2017, 05, 18), new DateTime(2018, 05, 10), new DateTime(2018, 02, 09), 85.54d, null, null, null, null, 511210, null, null, null);
```
Comment (hunterfries) [inline src/IQFeed.CSharpApiClient.Tests/Streaming/Level1/Messages/FundamentalMessageTests.cs:21]: Same thing here

File: src/IQFeed.CSharpApiClient/Extensions/DictionaryExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            return ToString(dictionary, (key, value) => $"{key.ToString()}: {value.ToString()}", separator);
+        }
+
+        public static string ToString<TK, TV>(this IDictionary<TK, TV> dictionary, Func<TK, TV, string> formatFunc, string seperator)
+        {
+            var output = new StringBuilder();
+            foreach (var keyValuePair in dictionary)
+            {
+                if (output.Length > 0)
+                    output.Append(seperator);
```
Comment (hunterfries) [inline src/IQFeed.CSharpApiClient/Extensions/DictionaryExtensions.cs:22]: I'm a proponent of wrapping brackets around one liners... @mathpaquette thoughts on style here?

Comment (hunterfries) [reply src/IQFeed.CSharpApiClient/Extensions/DictionaryExtensions.cs:22]: Just gonna make this the placeholder for this in the rest of the file.

File: src/IQFeed.CSharpApiClient/Lookup/MarketSummary/Messages/MarketSummaryMessage.cs [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+                        break;
+
+                    default:
+                        var expr2 = (UnaryExpression)property.Body;
+                        var prop2 = (PropertyInfo)((MemberExpression)expr2.Operand).Member;
+                        output.Append(ToStringNameValue(prop2.Name, prop2.GetValue(this)?.ToString()));
+                        break;
+                }
+
+                //var expr = (DynamicExpression)property.Body;
```
Comment (hunterfries) [inline src/IQFeed.CSharpApiClient/Lookup/MarketSummary/Messages/MarketSummaryMessage.cs:545]: What do we want to do with this commented code?

File: src/IQFeed.CSharpApiClient/Streaming/Level1/DynamicFieldset.cs [hunk 5]
```diff
@@ -133,12 +133,18 @@ public enum DynamicFieldset
...
         MostRecentTrade,
 
+        [FieldsetDescription("Most Recent Trade Aggressor", typeof(int))]
+        MostRecentTradeAggressor,
+
         [FieldsetDescription("Most Recent Trade Conditions", typeof(string))]
         MostRecentTradeConditions,
 
-        [FieldsetDescription("Most Recent Trade Date", typeof(DateTime))]
+        [FieldsetDescription("Most Recent Trade Date", typeof(DateTime), "MM/dd/yyyy")]
```
Comment (hunterfries) [inline src/IQFeed.CSharpApiClient/Streaming/Level1/DynamicFieldset.cs:142]: What's the goal of putting the attribute only this field of type DateTime?

**Comment by mathpaquette** [comment]
we need to rebase this one.

File: src/IQFeed.CSharpApiClient/Lookup/Chains/ChainsRequestFormatter.cs [hunk 6]
```diff
@@ -20,10 +20,12 @@ public string ReqChainFutureOption(string symbol, OptionSideFilterType optionSid
...
         public string ReqChainIndexEquityOption(string symbol, OptionSideFilterType optionSideFilter, string monthCodes, int? nearMonths = null, BinaryOptionFilterType binaryOptionFilter = BinaryOptionFilterType.Include,
-            OptionFilterType optionFilter = OptionFilterType.None, int? filterValue1 = null, int? filterValue2 = null, string requestId = null)
+            OptionFilterType optionFilter = OptionFilterType.None, int? filterValue1 = null, int? filterValue2 = null, string requestId = null, bool includeNonStandardOptions = false)
         {
-            var request = $"CEO,{symbol},{optionSideFilter.ToString().ToLower()},{monthCodes},{nearMonths},{(int)binaryOptionFilter},{(int)optionFilter},{filterValue1},{filterValue2},{requestId}{IQFeedDefault.ProtocolTerminatingCharacters}";
+            var wireFormatIncludeNonStandardOptions = includeNonStandardOptions ? 1 : 0;
+            var request = $"CEO,{symbol},{optionSideFilter.ToString().ToLower()},{monthCodes},{nearMonths},{(int)binaryOptionFilter},{(int)optionFilter},{filterValue1},{filterValue2},{requestId},{wireFormatIncludeNonStandardOptions}{IQFeedDefault.ProtocolTerminatingCharacters}";
```
Comment (sharok) [inline src/IQFeed.CSharpApiClient/Lookup/Chains/ChainsRequestFormatter.cs:28]: I think `binaryOptionFilter` doesn't exist anymore. I don't see it in docs at least. Moreover, I didn't tried your branch, but I changed protocol to 6.1 ( in master code) and added `includeNonStandardOptions` and got syntax error response. After deleting `binaryOptionFilter` the error disappeared.

Comment (mathpaquette) [reply src/IQFeed.CSharpApiClient/Lookup/Chains/ChainsRequestFormatter.cs:28]: @sharok thanks

**Comment by mathpaquette** [comment]
@NichUK could you please rebase this PR?

**Comment by NichUK** [comment]
Sorry, I haven’t been around for a bit. Happy to look at this again.

**Comment by mathpaquette** [comment]
> Sorry, I haven’t been around for a bit. Happy to look at this again.

Hey @NichUK, it would be nice. I've received some comments that this PR is not following existing design patterns but I think this is due to the fact the previously we're supporting different types before going all-in for doubles. 

My recommendation would be to start with a rebase and Ill begin a review to see what changes are needed before merging it.

I know you put some effort into it so Ill prefer to prioritize your work because newer protocol support is getting bigger.

THanks,
Mathieu

**Comment by NichUK** [comment]
OK, so this is a no go. Between the removal of the generics, and the fact that we both implemented Dynamic Fields, but differently, rebasing this is just not going to happen.
So I've taken an up to date version, and I'm manually cherry-picking the protocol changes on to it. It's not too hard, so shouldn't take too long, and it'll be way cleaner.
I'm just doing 6.1 at this stage, and then we can overlay 6.2 on top.
Annoyingly, the current IQFeed beta, has a bug in it regarding OptionChains (the response is always in 6.2 dialect, which is different from previous, no matter what protocol you set - I've reported it), so I'll either have to downgrade for a bit, or suffer the annoyance of tests that fail. Shouldn't be long for a fix though, it's a simple issue for them to solve.
Give me a day or two, and I'll open a new PR for it.

**Comment by mathpaquette** [comment]
@NichUK makes sense. thank you!

**Pull request closed without merging**